File: server/EvalutionAlghoritm2/Fitness.js
import reviver from "./JSONReviver.js";

export default function Fitness(
  schedule,
  recommended_schedules,
  penaltySameRecSc,
  penaltyGrWin,
  penaltySameTimesSc,
  penaltyTeachWin
) {
  schedule = JSON.parse(schedule, reviver);
  for (let value of schedule.scheduleForGroups.values()) sortSchedule(value);
  for (let value of schedule.scheduleForTeachers.values()) sortSchedule(value);
  for (let value of schedule.scheduleForAudiences.values()) sortSchedule(value);
  let fitnessGr = fitnessByGroups(
    schedule.scheduleForGroups,

    penaltySameTimesSc,
    penaltyGrWin
  );
  let fitnessTeach = fitnessByTeachers(
    schedule.scheduleForTeachers,
    penaltySameTimesSc,
    penaltyTeachWin
  );
  let fitnessAud =
    penaltySameTimesSc === 0
      ? 0
      : fitnessByAudiences(schedule.scheduleForAudiences, penaltySameTimesSc);
  let fitnessSameRecSc =
    penaltySameRecSc === 0
      ? 0
      : fitnessSameSchedules(
          schedule.scheduleForAudiences,
          recommended_schedules,
          penaltySameRecSc
        );
  let fitnessValue =
    fitnessGr.fitnessValue +
    fitnessTeach.fitnessValue +
    fitnessAud.fitnessValue +
    fitnessSameRecSc;

  return {
    fitnessValue,
    fitnessGr,
    fitnessTeach,
    fitnessAud,
    fitnessSameRecSc,
  };
}

function sortSchedule(schedule) {
  schedule.sort(function (a, b) {
    if (a.day_week > b.day_week) return 1;
    if (a.day_week < b.day_week) return -1;
    if (a.number_pair > b.number_pair) return 1;
    if (a.number_pair < b.number_pair) return -1;
    if (a.pair_type > b.pair_type) return 1;
    if (a.pair_type < b.pair_type) return -1;
    return 0;
  });
}
function fitnessByGroups(schedule, penaltySameTimesSc, penaltyGrWin) {
  let fitnessGrWin = 0;
  let fitnessSameTimesSc = 0;
  for (let value of schedule.values()) {
    // fitnessGrWin +=
    //   penaltyGrWin === 0 ? 0 : fitnessWindows(value, penaltyGrWin);
    // fitnessSameTimesSc +=
    //   penaltySameTimesSc === 0
    //     ? 0
    //     : fitnessSameTimes(value, penaltySameTimesSc);
    let { fitnessSameTimes, fitnessWindows } = FitnessSameTimesAndWindows(
      value,
      penaltySameTimesSc,
      penaltyGrWin
    );
    fitnessGrWin += fitnessWindows;
    fitnessSameTimesSc += fitnessSameTimes;
  }
  let fitnessValue = fitnessGrWin + fitnessSameTimesSc;
  return {
    fitnessGrWin,
    fitnessSameTimesSc,
    fitnessValue,
  };
}
function fitnessByTeachers(schedule, penaltySameTimesSc, penaltyTeachWin) {
  let fitnessTeachWin = 0;
  let fitnessSameTimesSc = 0;
  for (let value of schedule.values()) {
    // fitnessTeachWin +=
    //   penaltyTeachWin === 0 ? 0 : fitnessWindows(value, penaltyTeachWin);

    // fitnessSameTimesSc +=
    //   penaltySameTimesSc === 0
    //     ? 0
    //     : fitnessSameTimes(value, penaltySameTimesSc);
    let { fitnessSameTimes, fitnessWindows } = FitnessSameTimesAndWindows(
      value,
      penaltySameTimesSc,
      penaltyTeachWin
    );
    fitnessTeachWin += fitnessWindows;
    fitnessSameTimesSc += fitnessSameTimes;
  }
  let fitnessValue = fitnessTeachWin + fitnessSameTimesSc;
  return {
    fitnessTeachWin,
    fitnessSameTimesSc,
    fitnessValue,
  };
}
function fitnessByAudiences(schedule, penaltySameTimesSc) {
  let fitnessSameTimesSc = 0;
  for (let value of schedule.values()) {
    fitnessSameTimesSc +=
      penaltySameTimesSc === 0
        ? 0
        : fitnessSameTimes(value, penaltySameTimesSc);
  }
  let fitnessValue = fitnessSameTimesSc;
  return {
    fitnessSameTimesSc,
    fitnessValue,
  };
}
function getClosePair(array) {
  let max = array[0];
  for (let i = 1; i < array.length; i++) {
    if (array[i] && max) {
      if (max.number_pair < array[i]?.number_pair) max = array[i];
    } else if (!max) max = array[i];
  }
  return max;
}

// function FitnessSameTimesAndWindows(schedule, penaltySameTimesSc, penaltyWin) {
//   let fitnessSameTimes = 0;
//   let fitnessWindows = 0;
//   let currentDay = -1;
//   let lastTop;
//   let lastBot;
//   let lastTotal;
//   for (let i = -1; i < schedule.length - 1; i++) {
//     // Переход на новый день
//     if (currentDay !== schedule[i + 1].day_week) {
//       currentDay = schedule[i + 1].day_week;
//       lastTop = null;
//       lastBot = null;
//       lastTotal = null;
//     }
//     // Если нету перехода на новый день
//     else {
//       if (schedule[i].pair_type === 1) {
//         lastTop = schedule[i];
//       }
//       else if (schedule[i].pair_type === 2) {
//         lastBot = schedule[i];
//       }
//       else if (schedule[i].pair_type === 3) {
//         lastTotal = schedule[i];
//       }
//       // Проверка накладки
//       if (schedule[i + 1].number_pair === schedule[i].number_pair) {
//         // Если пара общая а какая либо в эту пару числитель/знаменатель или одинаковый тип
//         if (schedule[i + 1].pair_type === 1 && (lastTop?.number_pair === schedule[i + 1].number_pair || lastTotal?.number_pair === schedule[i + 1].number_pair)) {
//           fitnessSameTimes += penaltySameTimesSc;
//           continue;
//         }
//         if (schedule[i + 1].pair_type === 2 && (lastBot?.number_pair === schedule[i + 1].number_pair || lastTotal?.number_pair === schedule[i + 1].number_pair)) {
//           fitnessSameTimes += penaltySameTimesSc;
//           continue;
//         }
//         if (schedule[i + 1].pair_type === 3 && (lastTop?.number_pair === schedule[i + 1].number_pair ||
//           lastBot?.number_pair === schedule[i + 1].number_pair || lastTotal?.number_pair === schedule[i + 1].number_pair)) {
//           fitnessSameTimes += penaltySameTimesSc;
//           continue;
//         }
//       }
//       // Проверка на окна
//       else {
//         // Если явное окно
//         if (schedule[i + 1].number_pair - schedule[i].number_pair > 1) {
//           fitnessWindows += (schedule[i + 1].number_pair - schedule[i].number_pair - 1) * penaltyWin;
//           continue;
//         }
//         // Если числитель/общая а перед ним знаменатель
//         if ((schedule[i + 1].pair_type === 1 || schedule[i + 1].pair_type === 3) && lastBot?.number_pair === schedule[i + 1].number_pair - 1) {
//           if (lastTop?.number_pair < schedule[i + 1].number_pair - 1) {
//             fitnessWindows += (schedule[i + 1].number_pair - lastTop?.number_pair - 1) * penaltyWin;
//             continue;
//           }
//           if (lastTotal?.number_pair < schedule[i + 1].number_pair - 1) {
//             fitnessWindows += (schedule[i + 1].number_pair - lastTotal?.number_pair - 1) * penaltyWin;
//             continue;
//           }
//         }
//         // Если знаменатель/общая а перед ним числитель
//         if ((schedule[i + 1].pair_type === 2 || schedule[i + 1].pair_type === 3)) {
//           if (lastTop?.number_pair === schedule[i + 1].number_pair - 1) {
//             if (lastBot?.number_pair < schedule[i + 1].number_pair - 1) {
//               fitnessWindows += (schedule[i + 1].number_pair - lastBot?.number_pair - 1) * penaltyWin;
//               continue;
//             }
//             if (lastTotal?.number_pair < schedule[i + 1].number_pair - 1) {
//               fitnessWindows += (schedule[i + 1].number_pair - lastTotal?.number_pair - 1) * penaltyWin;
//               continue;
//             }
//           }
//           if (schedule.find(sc => sc.day_week === currentDay && sc.number_pair === schedule[i + 1].number_pair - 1 && sc.pair_type === 1)) {
//             if (lastBot?.number_pair < schedule[i + 1].number_pair - 1) {
//               fitnessWindows += (schedule[i + 1].number_pair - lastBot?.number_pair - 1) * penaltyWin;
//               continue;
//             }
//             if (lastTotal?.number_pair < schedule[i + 1].number_pair - 1) {
//               fitnessWindows += (schedule[i + 1].number_pair - lastTotal?.number_pair - 1) * penaltyWin;
//               continue;
//             }
//           }
//         }
//       }
//     }
//   }
//   return { fitnessWindows, fitnessSameTimes }
// }

function FitnessSameTimesAndWindows(schedule, penaltySameTimesSc, penaltyWin) {
  let lastTop = null;
  let lastBot = null;
  let lastTotal = null;
  let fitnessWindows = 0;

  let fitnessSameTimes = 0;

  for (let i = 0; i < schedule.length - 1; i++) {
    if (+schedule.id_class === 2498) {
      console.log();
      console.log();
    }
    if (schedule[i].day_week === schedule[i + 1].day_week) {
      //накладки
      if (schedule[i].number_pair === schedule[i + 1].number_pair) {
        //все случае кроме если стоит числитель и знаменатель
        if (
          schedule[i].pair_type === schedule[i + 1].pair_type ||
          schedule[i].pair_type === 3 ||
          schedule[i + 1].pair_type === 3
        )
          fitnessSameTimes += penaltySameTimesSc;
      } else {
        switch (schedule[i].pair_type) {
          case 1:
            lastTop = schedule[i];
            break;
          case 2:
            lastBot = schedule[i];
            break;
          case 3:
            lastTotal = schedule[i];
            break;
        }
        let closePair =
          schedule[i + 1].pair_type === 3
            ? [
                getClosePair([lastTop, lastTotal]),
                getClosePair([lastBot, lastTotal]),
              ]
            : schedule[i + 1].pair_type === 2
            ? getClosePair([lastBot, lastTotal])
            : getClosePair([lastTop, lastTotal]);
        if (closePair)
          if (closePair?.length) {
            let closePair1 = closePair[0];
            let closePair2 = closePair[1];
            if (
              closePair1 &&
              closePair2 &&
              closePair1?.pair_type !== closePair2?.pair_type
            ) {
              /* if (
                schedule[i + 1].number_pair - closePair1.number_pair - 1 >
                0
              ) {
                console.log();
                console.log();
              }
              if (
                schedule[i + 1].number_pair - closePair2.number_pair - 1 >
                0
              ) {
                console.log();
                console.log();
              }*/
              fitnessWindows +=
                (schedule[i + 1].number_pair - closePair1.number_pair - 1) *
                penaltyWin;
              fitnessWindows +=
                (schedule[i + 1].number_pair - closePair2.number_pair - 1) *
                penaltyWin;
            } else {
              if (closePair1) {
                /* if (
                  schedule[i + 1].number_pair - closePair1.number_pair - 1 >
                  0
                ) {
                  console.log();
                  console.log();
                }*/
                fitnessWindows +=
                  (schedule[i + 1].number_pair - closePair1.number_pair - 1) *
                  penaltyWin;
              } else if (closePair2) {
                /* if (
                  schedule[i + 1].number_pair - closePair2.number_pair - 1 >
                  0
                ) {
                  console.log();
                  console.log();
                }*/

                fitnessWindows +=
                  (schedule[i + 1].number_pair - closePair2.number_pair - 1) *
                  penaltyWin;
              }
            }
          } else {
            /*if (schedule[i + 1].number_pair - closePair.number_pair - 1 > 0) {
              console.log();
              console.log();
            }*/
            fitnessWindows +=
              (schedule[i + 1].number_pair - closePair.number_pair - 1) *
              penaltyWin;
          }
      }
    } else {
      lastTop = null;
      lastBot = null;
      lastTotal = null;
    }
  }
  return { fitnessWindows, fitnessSameTimes };
}

function fitnessWindows(schedule, penaltyWin) {
  let fitnessValue = 0;
  let len = schedule.length;
  for (let i = 0; i < len - 1; i++) {
    for (let j = i + 1; j < len; j++) {
      if (+schedule[i].day_week === +schedule[j].day_week) {
        if (+schedule[i].number_pair !== +schedule[j].number_pair) {
          if (schedule[i].pair_type === 3) {
            while (j < len && schedule[i].day_week === +schedule[j].day_week) {
              if (
                schedule[j - 1].pair_type !== schedule[j].pair_type &&
                schedule[j].pair_type !== 3
              )
                fitnessValue +=
                  (schedule[j].number_pair - schedule[i].number_pair - 1) *
                  penaltyWin;
              else break;
              j += 1;
            }
            break;
          }

          fitnessValue +=
            (schedule[j].number_pair - schedule[i].number_pair - 1) *
            penaltyWin;
          /*if (schedule[j].number_pair - schedule[i].number_pair - 1 > 0) {
            console.log();
            console.log();
          }*/
        }
        if (
          +schedule[i].pair_type === +schedule[j].pair_type ||
          +schedule[j].pair_type === 3
        )
          break;
      } else break;
    }
  }
  return fitnessValue;
}
function fitnessSameTimes(schedule, penaltySameTimesSc) {
  let fitnessValue = 0;
  let i = 0;
  while (i < schedule.length - 1) {
    //в границе одного дня
    if (schedule[i].day_week === schedule[i + 1].day_week) {
      //если у двух пар совпадает номер пары
      if (schedule[i].number_pair === schedule[i + 1].number_pair) {
        //все случае кроме если стоит числитель и знаменатель
        if (
          schedule[i].pair_type === schedule[i + 1].pair_type ||
          schedule[i].pair_type === 3 ||
          schedule[i + 1].pair_type === 3
        )
          fitnessValue += penaltySameTimesSc;
      }
    }
    i++;
  }
  return fitnessValue;
}

function fitnessSameSchedules(
  schedule,
  recommended_schedules,
  penaltySameRecSchedules
) {
  let fitnessSameRecSc = 0;
  for (let value of schedule.values()) {
    for (let clas of value) {
      let recSc = recommended_schedules.filter(
        (rs) => +rs.id_class === +clas.id_class
      );
      if (recSc.length) {
        let isSame = recSc.filter(
          (rs) =>
            +rs.day_week === +clas.day_week &&
            +rs.number_pair === +clas.number_pair
        );
        if (!isSame.length) fitnessSameRecSc += penaltySameRecSchedules;
      }
    }
  }
  return fitnessSameRecSc;
}

File: server/EvalutionAlgoritmCpp/EvalutionAlgoritmCpp/EvulationAlgoritmCpp/Fitness.cpp
#include "Fitness.h"
#include <map>
#include <iostream>
#include <vector>
#include "SortSchedule.h"
using namespace std;
map<string, double> FitnessByGroups(map <int, vector<schedule>> &scheduleForGroups,const int &max_day, const double &penaltySameTimesSc, const double &penaltyGrWin);
map<string, double> FitnessByTeachers(map <int, vector<schedule>> &scheduleForTeachers,const int &max_day,const double &penaltySameTimesSc,const double &penaltyTeachWin);
map<string, double> FitnessByAudiences(map <int, vector<schedule>> &scheduleForAudiences,const double &penaltySameTimesSc);
double FitnessWindows(vector<schedule > i_schedule,const int &max_day,const double &penaltyWin);
double FitnessSameTimes(vector<schedule> i_schedule,const double &penaltySameTimesSc);
double FitnessSameSchedules(map <int, vector<schedule>>&i_schedule, const vector < recommended_schedule> &recommedned_schedules,const double &penaltySameRecSc);

void Fitness(individ& i_schedule, const vector <recommended_schedule>& recommended_schedules, const int& max_day, const double& penaltySameRecSc, const double& penaltyGrWin, const double& penaltySameTimesSc, const double& penaltyTeachWin){
    bool sortedGroups = false,sortedTeachers= false,sortedAudiences=false;
    auto itGr=i_schedule.scheduleForGroups.begin(),itTeach= i_schedule.scheduleForTeachers.begin(),itAud= i_schedule.scheduleForAudiences.begin();
    while(!sortedGroups || !sortedTeachers || !sortedAudiences){
        if(itGr!= i_schedule.scheduleForGroups.end()){
            SortSchedule(itGr->second);
            ++itGr;
        }
        else
            sortedGroups=true;
        if(itTeach!= i_schedule.scheduleForTeachers.end()){
            SortSchedule(itTeach->second);
            ++itTeach;
        }
        else
            sortedTeachers=true;
        if(itAud!= i_schedule.scheduleForAudiences.end()){
            SortSchedule(itAud->second);
            ++itAud;
        }
        else
            sortedAudiences=true;

    }
    map<string, double> fitnessGr=FitnessByGroups(i_schedule.scheduleForGroups,max_day,penaltySameTimesSc,penaltyGrWin);
    map<string, double> fitnessTeach = FitnessByTeachers(i_schedule.scheduleForTeachers,max_day,penaltySameTimesSc,penaltyTeachWin);
    map<string, double> empty = { {"fitnessValue",0},{"fitnessSameTimesSc",0}};
    map<string, double> fitnessAud = penaltySameTimesSc == 0 ? empty : FitnessByAudiences(i_schedule.scheduleForAudiences,penaltySameTimesSc);
    double fitnessSameRecSc = FitnessSameSchedules(i_schedule.scheduleForAudiences,recommended_schedules,penaltySameRecSc);

    double fitnessValue = fitnessGr["fitnessValue"] + fitnessTeach["fitnessValue"] + fitnessAud["fitnessValue"] + fitnessSameRecSc;
    i_schedule.fitnessValue = fitness(fitnessValue, fitnessGr, fitnessTeach, fitnessAud, fitnessSameRecSc);
}


map<string, double> FitnessByGroups(map <int, vector<schedule>>& scheduleForGroups, const int& max_day, const double& penaltySameTimesSc, const double& penaltyGrWin){
    double fitnessGrWin = 0;
    double fitnessSameTimesSc = 0;
    auto itGr=scheduleForGroups.begin();
    while(itGr!=scheduleForGroups.end()){
        fitnessGrWin += penaltyGrWin == 0 ? 0 : FitnessWindows(itGr->second,max_day,penaltyGrWin);
        fitnessSameTimesSc += penaltySameTimesSc == 0 ? 0 : FitnessSameTimes(itGr->second,penaltySameTimesSc);
        ++itGr;
    }
    double fitnessValue = fitnessGrWin + fitnessSameTimesSc;
    map<string, double> fitnessV;
    fitnessV["fitnessValue"] = fitnessValue;
    fitnessV["fitnessGrWin"] = fitnessGrWin;
    fitnessV["fitnessSameTimesSc"] = fitnessSameTimesSc;
    return fitnessV;
}
map<string, double> FitnessByTeachers(map <int, vector<schedule>>& scheduleForTeachers, const int& max_day, const double& penaltySameTimesSc, const double& penaltyTeachWin){
    double fitnessTeachWin = 0;
    double fitnessSameTimesSc = 0;
    auto itTeach=scheduleForTeachers.begin();
    while(itTeach!=scheduleForTeachers.end()){
        fitnessTeachWin += penaltyTeachWin == 0 ? 0 : FitnessWindows(itTeach->second,max_day,penaltyTeachWin);
        fitnessSameTimesSc += penaltySameTimesSc == 0 ? 0 : FitnessSameTimes(itTeach->second,penaltySameTimesSc);
        ++itTeach;
    }
    double fitnessValue = fitnessTeachWin + fitnessSameTimesSc;
    map<string, double> fitnessV;
    fitnessV["fitnessValue"] = fitnessValue;
    fitnessV["fitnessTeachWin"] = fitnessTeachWin;
    fitnessV["fitnessSameTimesSc"] = fitnessSameTimesSc;
    return fitnessV;
}

map<string, double> FitnessByAudiences(map <int, vector<schedule>>&scheduleForAudiences, const double& penaltySameTimesSc){
    double fitnessSameTimesSc = 0;
    auto itAud=scheduleForAudiences.begin();
    while(itAud!=scheduleForAudiences.end()){
        fitnessSameTimesSc += penaltySameTimesSc == 0 ? 0 : FitnessSameTimes(itAud->second,penaltySameTimesSc);
        ++itAud;
    }
    double fitnessValue = fitnessSameTimesSc;
    map<string, double> fitnessV;
    fitnessV["fitnessValue"] = fitnessValue;
    fitnessV["fitnessSameTimesSc"] = fitnessSameTimesSc;
    return fitnessV;
}


double FitnessWindows(vector<schedule > i_schedule, const int& max_day, const double& penaltyWin){
    double fitnessWindows = 0;
    for(int current_day = 1; current_day <=max_day;current_day++){

        vector<schedule> schedule_top;
        vector<schedule> schedule_bot;
        for(schedule &clas: i_schedule){
            if(clas.day_week == current_day){
                switch(clas.pair_type){
                    case 1: schedule_top.push_back(clas);
                    break;
                    case 2: schedule_bot.push_back(clas);
                    break;
                    case 3: schedule_bot.push_back(clas);
                            schedule_top.push_back(clas);
                            break;
                }
            }
        }

        vector<vector<schedule>> arr;

        if(!schedule_top.empty())
        for(size_t i = 0; i<schedule_top.size()-1;i++){
            int diff = schedule_top[i+1].number_pair-schedule_top[i].number_pair -1;
            if(diff>1){
                fitnessWindows+=diff*penaltyWin;
                if(schedule_top[i+1].pair_type==3 && schedule_top[i].pair_type==3){
                    vector<schedule> arrT;
                    arrT.push_back(schedule_top[i]);
                    arrT.push_back(schedule_top[i+1]);
                    arr.push_back(arrT);
                }

            }

        }
        if (!schedule_bot.empty())
        for(size_t i = 0; i<schedule_bot.size()-1;i++){
            int diff = schedule_bot[i+1].number_pair-schedule_bot[i].number_pair-1;
            if(diff>1){
                if(schedule_bot[i+1].pair_type==3 && schedule_bot[i].pair_type ==3){
                    bool flag = false;
                    for (vector<schedule> arrJ : arr) {
                        if (arrJ[0].compare(schedule_bot[i]) && arrJ[1].compare(schedule_bot[i + 1])) {
                            continue;
                        }
                    }
                }

                fitnessWindows+=diff*penaltyWin;
            }
        }
    }
    return fitnessWindows;
}

double FitnessSameSchedules(map <int, vector<schedule>>& i_schedule, const vector < recommended_schedule>& recommedned_schedules, const double& penaltySameRecSc){
    double fitnessSameRecSc = 0;
     auto itSc=i_schedule.begin();
     while(itSc!=i_schedule.end()){
         for(const schedule &clas : itSc->second){
             vector<recommended_schedule> recSc;
             for(recommended_schedule rs: recommedned_schedules){
                 if(rs.id_class ==  clas.id_class)
                     recSc.push_back(rs);
             }
             if(!recSc.empty()){
                 vector<recommended_schedule> isSame;
                 for(recommended_schedule rs: recSc){
                     if(rs.day_week == clas.day_week && rs.number_pair == clas.number_pair)
                         isSame.push_back(rs);
                 }
                 if(!isSame.empty())
                     fitnessSameRecSc+=penaltySameRecSc;
             }
         }


         ++itSc;
    }
     return fitnessSameRecSc;

}

double FitnessSameTimes(vector<schedule > i_schedule, const double& penaltySameTimesSc){
    double fitnessSameTimes = 0;
    schedule lastTop,lastBot,lastTotal;
    int cur_day = -1;
    for(size_t i=-1;i<i_schedule.size()-1;i++){
        if(i_schedule[i+1].day_week != cur_day){
            cur_day = i_schedule[i+1].day_week;
            lastTop = schedule();
            lastBot = schedule();
            lastTotal = schedule();
            continue;
        }
        switch (i_schedule[i].pair_type)
        {
            case 1:
                    lastTop = i_schedule[i];
                    break;
            case 2:
                    lastBot = i_schedule[i];
                    break;
            case 3:
                    lastTotal = i_schedule[i];
                    break;
        }
        if(i_schedule[i+1].number_pair == i_schedule[i].number_pair){
            if(i_schedule[i+1].pair_type == 1 && (lastTop.isNullOrEmpty() ? false : lastTop.number_pair ==i_schedule[i+1].number_pair || lastTotal.isNullOrEmpty() ? false : lastTotal.number_pair ==i_schedule[i+1].number_pair)){
                fitnessSameTimes+=penaltySameTimesSc;
                continue;
            }
            if(i_schedule[i+1].pair_type == 2 && (lastBot.isNullOrEmpty() ? false : lastBot.number_pair ==i_schedule[i+1].number_pair || lastTotal.isNullOrEmpty() ? false : lastTotal.number_pair ==i_schedule[i+1].number_pair)){
                fitnessSameTimes+=penaltySameTimesSc;
                continue;
            }
            if(i_schedule[i+1].pair_type == 3 && (lastTop.isNullOrEmpty() ? false : lastTop.number_pair ==i_schedule[i+1].number_pair || lastBot.isNullOrEmpty() ? false :lastBot.number_pair ==i_schedule[i+1].number_pair || lastTotal.isNullOrEmpty() ? false : lastTotal.number_pair ==i_schedule[i+1].number_pair)){
                fitnessSameTimes+=penaltySameTimesSc;
                continue;
            }
        }
    }
    return fitnessSameTimes;
}



